### PR TITLE
Python connector CDF reads kernel integration

### PR DESCRIPTION
**Pull request by PatrickJin-db**
Main differences compared to the prototype #609 include:
* Not using a prototype branch of delta-kernel-rs
* Better code organization
* Ensuring that files are closed and tempfiles are cleaned up even upon failure
* Fixed some of the tests

Regarding performance, most of the time is spent in kernel, and writing the delta log is negligible in comparison.
image

### CODE REVIEW

File: python/delta-kernel-rust-sharing-wrapper/Cargo.toml
Comment (OussamaSaoudi-db) [inline python/delta-kernel-rust-sharing-wrapper/Cargo.toml:14]:
```suggestion
delta_kernel = { git = "https://github.com/delta-io/delta-kernel-rs.git", rev = "be1453f", features = ["cloud", "default", "default-engine"]}
```

File: python/delta-kernel-rust-sharing-wrapper/src/lib.rs
Comment (OussamaSaoudi-db) [inline python/delta-kernel-rust-sharing-wrapper/src/lib.rs:4]: Note: Both arrow and kernel have their own `SchemaRef`. This is up to personal preference, but you may want to give this an alias so that the type of `SchemaRef` doesn't become confusing.
```suggestion
use arrow::datatypes::SchemaRef as ArrowSchemaRef;
```

Comment (OussamaSaoudi-db) [reply python/delta-kernel-rust-sharing-wrapper/src/lib.rs:4]: You may also want to do the same for `TableChangesScan`:
```rust
use delta_kernel::table_changes::scan::TableChangesScan as KernelTableChangesScan
```

**Review by OussamaSaoudi-db — COMMENTED** [review]
Most pressing issue is updating to the main branch.

File: python/delta_sharing/delta_sharing.py
Comment (linzhou-db) [reply python/delta_sharing/delta_sharing.py:245]: I see why, let's add a comment to this line about it.

File: python/delta_sharing/rest_client.py
Comment (linzhou-db) [inline python/delta_sharing/rest_client.py:437]: and let's try some logs for the ease of debugging in the future when issues happen.
Such as printing out # of lines received from the server.

File: python/delta_sharing/reader.py
Comment (PatrickJin-db) [reply python/delta_sharing/reader.py:274]: One case could be if only metadata changes occurred during the queried version range.

File: python/delta-kernel-rust-sharing-wrapper/src/lib.rs
Comment (PatrickJin-db) [reply python/delta-kernel-rust-sharing-wrapper/src/lib.rs]: I think `ArrowSchemaRef` makes sense, but `TableChangesScan` is only used once so I don't think using an alias for it is as beneficial.

File: python/delta_sharing/delta_sharing.py
Comment (PatrickJin-db) [reply python/delta_sharing/delta_sharing.py:247]: done, PTAL

Comment (linzhou-db) [reply python/delta_sharing/delta_sharing.py:247]: looks good!

**Review by linzhou-db — APPROVED** [review]
The python code changes look good! 
Please get approval for the rust code before merging! 
And could you also check the failure in the wheel build jobs and see if it's possible to fix them?

**Pull request merged**
